### DISCUSSION TITLE
use SDK 4 for Model AXIS P3925-R

### DISCUSSION THREAD

**Discussion by axromero** [AxisCommunications/discussions#230] [ACAP] [answered]
fatal error: axsdk/axparameter.h: No such file or directory
#include <axsdk/axparameter.h>

My goal is to use SDK 4 for Model AXIS P3925-R.
But i can not find axparameter.h in the container ?

DOCKER IMAGE:
ARG ARCH=armv7hf
ARG VERSION=1.7
ARG UBUNTU_VERSION=22.04
ARG REPO=axisecp
ARG SDK=acap-native-sdk

FROM ${REPO}/${SDK}:${VERSION}-${ARCH}-ubuntu${UBUNTU_VERSION}

https://axiscommunications.github.io/acap-documentation/docs/api/native-sdk-api.html
https://www.axis.com/techsup/developer_doc/acap3/3.5/api/axparameter/html/axparameterexample_8c-example.html

**Comment by onie5** [comment]
ARG VERSION=3.5
ARG UBUNTU_VERSION=20.04
ARG SDK=acap-sdk

**Reply by axromero** [reply]
Thank you i am using now:
https://github.com/AxisCommunications/acap3-sdk/tree/main

# Repository and tag variables
ARG ARCH=armv7hf
ARG API_VERSION=3.5
ARG TOOLCHAIN_VERSION=3.5
ARG TOOLCHAINS_UBUNTU_VERSION=20.04
ARG SDK_UBUNTU_VERSION=22.04
ARG REPO=axisecp

There is all what is missing in SDK 4 .

But i still have some issues like you with axparameter.
In my case it is that the ax_parameter_register_callback() is not working ?
I can poll via ax_parameter_get.

**Reply by onie5** [reply]
Can you explain a bit why you say it is not working? Maybe your arguments are not correct. A short snap of your code may be helpful.

The parameter library is very tricky if not buggy, sometimes you may need to workaround it.

**Reply by axromero** [reply]
[AxParamTest-3.5SDK.zip](https://github.com/AxisCommunications/discussions/files/11431959/AxParamTest-3.5SDK.zip)

ParamHandler::handleParam1(const char *name, const char *value, gpointer data)
will never be called

LOG_CRIT check the /var/log/cr...

**Reply by onie5** [reply, marked as answer]
I see that you did not use the GMainLoop. This event loop monitors the parameter changes and executes the callback.

Adding this, your callback worked but you need to change your while-loop function to maybe a g_timeout_add function instead.

**Comment by axromero** [comment]
Thank you very much. It works the way you told :-)

But now i got the next issue ...
How do i stop the plugin properly ?
2023-05-10T11:26:09.771+02:00 axis-b8a44f008f22 [ WARNING ] acap-startstop: Application paramtest[15467] can not be stopped

**Reply by onie5** [reply]
Not sure what is causing this behavior with your code. Try re-working your event loop.

**Comment by axromero** [comment]
I think there was many reasons :-)
One was in the manifest.json the "runMode": "respawn",

Now every thing looks fine 

THANK YOU
